### DISCUSSION TITLE
Apply attributes to the block of a footnote?

### DISCUSSION THREAD

**Discussion by sjml** [jgm/djot#304] [Q&A]
The documentation says the footnote itself is a block element, but I can't figure out how to apply an attribute that will get applied to the rendered `<li>` in the HTML output. 

```djot
Here is a thing with a footnote.[^fn1]

[^fn1]: {key=val}
  This `<p>` element will have the attribute. 

  Sadly, this one won't.
```

I may just be pushing too hard to get the system to do something it's not wanting to do, but is there any good way to apply that attribute to the whole footnote?

I've settled on this in the meanwhile, but I don't love it. 

```djot
Here is a thing with a footnote.[^fn1]


[^fn1]: {key=val}
  :::
  This `<p>` is wrapped in a `<div>` that has the attribute now! 

  This one is in the same `<div>` so maybe it's good enough?
  :::
```

**Comment by Omikhleia** [comment]
```
footnote call[^1]

{keyfn=value}
[^1]:
  {keypara=value}
  Note text
```

Yields (https://djot.net/playground/) the following AST (abstract syntax tree):

```
doc
  para
    str text="footnote call"
    footnote_reference text="1"
footnotes
  ["1"] =
    footnote label="1" keyfn="value"
      para keypara="value"
        str text="Note text"
```

So as far as Djot input is concerned, it seems to me that the block attribute (here "keyfn=value") is correctly parsed (it must be on its own line preceding the block), and is where it should be in the AST.

However, it's not rendered in the HTML output (whether on the `<ol>`  or  `<li>` tag) -- So that's actually an issue in the Djot HTML renderer(s).

For the record, in my own SILE-based renderer, I use such block attributes for "styling" footnotes with a custom mark, e.g.

```
Foonote call[^my-note] ...

{mark="†"}
[^my-note]: Some note with a "dagger" reference mark ...
```

So your issue here really seems to be on the HTML renderer's side of things -- It might be worth reporting an issue on djot.lua and/or djot.js -- and Pandoc 3.1.13 also ignores the block attribute...

**Reply by vassudanagunta** [reply]
> I use such block attributes for "styling" footnotes with a custom mark

In the spirit of lightweight markup, the following would be cleaner, and would be supported by
a smart renderer that recognized the footnote label as also the recommended footnote mark. 

```
Foonote call[^†] ...

[^†]: Some note with a "dagger" reference mark ...
```

**Reply by Omikhleia** [reply]
@vassudanagunta 
Not necessarily:
- `[^my-note]` wouldn't mean alphabetic notes are intended, obviously... so it's probably not that a good an idea to conflate the note reference label and its styling...
- ... and I could have several notes marked with a symbol (and possibly the same † on different pages or in different chapters), but a note reference then has to be unique.

(But this is, anyway, outside the scope of the initial question here -- I just wanted to emphasize the the Djot input parser is not at stake here, and that the current problem lies in the HTML renderer, ignoring attributes on the footnote block)

**Reply by sjml** [reply]
Just checked and confirmed that Pandoc has the same issue with losing the attribute, but there it doesn't even seem to make it into the AST. @jgm, is this intentional behavior or a bug? The disparity in ASTs between djot.js and pandoc leads me to think it's just a yet-to-be-explored edge case. 

The Pandoc-produced AST from @Omikhleia's [example above](304#discussioncomment-9559906) is as follows:
```json
{
  "pandoc-api-version": [1, 23, 1],
  "meta": {},
  "blocks": [
    {
      "t": "Para",
      "c": [
        { "t": "Str", "c": "footnote call" },
        {
          "t": "Note",
          "c": [
            {
              "t": "Div",
              "c": [
                [
                  "",
                  [],
                  [
                    ["wrapper", "1"],
                    ["keypara", "value"]
                  ]
                ],
                [{ "t": "Para", "c": [{ "t": "Str", "c": "Note text" }] }]
              ]
            }
          ]
        }
      ]
    }
  ]
}
```

Djot.js, on the other hand, gives:
```json
{
  "tag": "doc",
  "references": {},
  "autoReferences": {},
  "footnotes": {
    "1": {
      "tag": "footnote",
      "label": "1",
      "children": [
        {
          "tag": "para",
          "children": [
            {
              "tag": "str",
              "text": "Note text"
            }
          ],
          "attributes": {
            "keypara": "value"
          }
        }
      ],
      "attributes": {
        "keyfn": "value"
      }
    }
  },
  "children": [
    {
      "tag": "para",
      "children": [
        {
          "tag": "str",
          "text": "footnote call"
        },
        {
          "tag": "footnote_reference",
          "text": "1"
        }
      ]
    }
  ]
}
```

(Interestingly, asking djot to output to Pandoc currently seems to drop all the attributes?) 
```json
{
  "pandoc-api-version": [
    1,
    23
  ],
  "meta": {},
  "blocks": [
    {
      "t": "Para",
      "c": [
        {
          "t": "Str",
          "c": "footnote"
        },
        {
          "t": "Space"
        },
        {
          "t": "Str",
          "c": "call"
        },
        {
          "t": "Note",
          "c": [
            {
              "t": "Para",
              "c": [
                {
                  "t": "Str",
                  "c": "Note"
                },
                {
                  "t": "Space"
                },
                {
                  "t": "Str",
                  "c": "text"
                }
              ]
            }
          ]
        }
      ]
    }
  ]
}
```

**Reply by jgm** [reply]
Pandoc doesn't have a slot for attributes in many elements, including Note.
We try to surround these with a Span or Div so the element isn't lost.

We should fix the djot HTML renderer so that it adds the attribute to the note; perhaps someone would like to open an issue for that?
